include/seastar/core/semaphore.hh

```diff
@@ -249,6 +249,26 @@ public:
     {
         static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
     }
+    basic_semaphore(basic_semaphore&& x) noexcept(std::is_nothrow_move_constructible_v<exception_factory>)
+        : exception_factory(std::move(x))
+        , _count(x._count)
+        , _wait_list(expiry_handler{*this})
+    {
+        // semaphore cannot be moved with non-empty waiting list
+        assert(x._wait_list.empty());
```

Comment:
You know what I think about assert()... This could have thrown instead of assert.
But, can't we change the waiting list's pointer, even if it's not empty?

Reply:
In 9b6e181e424fd18b780ade4df59e2c0c3a806742 we already restricted the move of the semaphore
after it was used (even if there were waiters in the past and they were already satisfied and no longer on the wait list).  The reason was to simplify move while there are waiters on the wait list.
So this patch follows the same lines.

Reply:
You can't throw in a noexecpt move constructor. And marking move constructors as throwing generally decreases stability since it adds exceptional paths that are never tested.

Reply:
Right.